doc/Language/101-basics.pod6

```diff
@@ -241,20 +241,33 @@ The program then counts the number of sets each player has won:
 %sets{$p2} += $r2;
 =end code
 
-The C<+=> assignment operator is a shortcut for:
+The above two statements involve the C<+=> compound assignment operator. They
+are a variant of the following two statements that use the simple assignment
+operator C<=> and that may be easier to understand at first sight:
 
 =begin code :preamble<my %sets;my $p1;my $p2;my $r1;my $r2>
 %sets{$p1} = %sets{$p1} + $r1;
 %sets{$p2} = %sets{$p2} + $r2;
 =end code
 
+For your program, the shorthand version using the C<+=> compound assignment
+operator is preferred over the longhand version using the simple assignment
+operator C<=>. This is so not only because the shorter version requires less
+typing, but also because the C<+=> operator silently initializes undefined
+values of the hash's key-value pairs to zero. In other words: by using C<+=>,
+there is no need to include a separate statement like C<%sets{$p1} = 0> before
+you can meaningfully add C<$r1> to it. We'll look at this is behavior in a bit
+more detail below.
+
+
 =head3 X<C<Any>| Any (Basics)> and X<C<+=>>
 
-C<+= $r1> means I<increase the value in the variable on the left by $r1>. In
-the first iteration C<%sets{$p1}> is not yet set, so it defaults to a special
-value called C<Any>. The addition and incrementing operators treat C<Any> as a
-number with the value of zero; the strings get automatically converted to
-numbers, as addition is a numeric operation.
+C<+= $r1> means I<increase the value in the variable on the left by $r1>. In the
+first iteration C<%sets{$p1}> is not yet defined, so it defaults to a special
+value called C<Any>. The C<+=> operator conveniently treats the undefined value
```

Comment:
I appreciate this was like this before but this could be a bit misleading about the type system and the role of type objects as the  "undefined value", it might be worth introducing a bit more detail about how `Any`  will become the zero or the empty string or whatever depending on the type RHS of the operator.

Reply:
I agree the current text isn't very instructive on the point of the type system, type objects and definedness. But I'm not sure how to clarify things in more detail without expanding the 101 too much, and possibly beyond its purpose. In addition I think your observation isn't necessarily related to issue # 2920, so perhaps it's better to deal with this point in a new issue.

The issue could, for instance, call attention to the fact that the 101 does not systematically introduce its reader to the Raku type system and the aforementioned concepts. I'd think that resolving the issue would require at least the following to be included to make these concepts understandable: that Raku is object oriented to the core; that everything in Raku is an object (or behaves like one), with objects being either "type objects" that define types, or "instances" that derive from type objects; that all objects have a property called definedness that indicates whether the object is an instance and represents a meaningful (initialized) value; that by default type objects are undefined while instances are defined; that (Any) is a type object that is the default value for i.a. hash values; and that compound assignment operators check whether their left operand is a type object/is undefined, and in case it is, implicitly initialize it with an instance having a value meaningful for the specific operator...

But even though this seems very useful info, it may be a bit too much for a 101 whose primary purpose seems to be to get the uninitiated up and running with Raku quickly. Or is the purpose of the 101 a different one?

Reply:
No, it's not. And it's likely that there's too much information there already. Just leave it this way.